src/banking_stage.rs

```diff
@@ -9,75 +9,102 @@ use entry::Entry;
 }
 
+impl Default for Config {
+    fn default() -> Config {
+        // TODO: Change this to Tick to enable PoH
+        Config::Sleep(Duration::from_millis(500))
+    }
+}
 impl BankingStage {
     /// Create the stage using `bank`. Exit when `verified_receiver` is dropped.
     /// Discard input packets using `packet_recycler` to minimize memory
     /// allocations in a previous stage such as the `fetch_stage`.
     pub fn new(
         bank: Arc<Bank>,
         verified_receiver: Receiver<Vec<(SharedPackets, Vec<u8>)>>,
-        tick_duration: Option<Duration>,
+        config: Config,
     ) -> (Self, Receiver<Vec<Entry>>) {
         let (entry_sender, entry_receiver) = channel();
-        let thread_hdl = Builder::new()
-            .name("solana-banking-stage".to_string())
+        let shared_verified_receiver = Arc::new(Mutex::new(verified_receiver));
+        let poh = PohRecorder::new(bank.clone(), entry_sender);
+        let tick_poh = poh.clone();
+
+        // Single thread to generate entries from many banks.
+        // This thread talks to poh_service and broadcasts the entries once they have been recorded.
+        // Once an entry has been recorded, its last_id is registered with the bank.
+        let tick_producer = Builder::new()
+            .name("solana-banking-stage-tick_producer".to_string())
             .spawn(move || {
-                let poh_service = PohService::new(bank.last_id(), tick_duration.is_some());
-
-                let mut last_tick = Instant::now();
-
-                loop {
-                    let timeout =
-                        tick_duration.map(|duration| duration - (Instant::now() - last_tick));
-
-                    if let Err(e) = Self::process_packets(
-                        timeout,
-                        &bank,
-                        &poh_service,
-                        &verified_receiver,
-                        &entry_sender,
-                    ) {
-                        match e {
-                            Error::RecvTimeoutError(RecvTimeoutError::Disconnected) => break,
-                            Error::RecvTimeoutError(RecvTimeoutError::Timeout) => (),
-                            Error::RecvError(_) => break,
-                            Error::SendError => break,
-                            _ => {
-                                error!("process_packets() {:?}", e);
-                                break;
-                            }
-                        }
-                    }
-                    if tick_duration.is_some() && last_tick.elapsed() > tick_duration.unwrap() {
-                        if let Err(e) = Self::tick(&poh_service, &entry_sender) {
-                            error!("tick() {:?}", e);
-                        }
-                        last_tick = Instant::now();
+                if let Err(e) = Self::tick_producer(tick_poh, config) {
+                    match e {
+                        Error::SendError => (),
```

Comment:
I'm confused about what causes this thread to exit, is it dropping entry_receiver?

Reply:
@rob-solana yea, is that ok?

Reply:
I guess.  I'm still trying to get used to how we manage threads, whose stacks end up being like the object state, and which I often feel unsure about how we clean 'em up.  Modal fullnodes require that we tear things down correctly, that we don't leak threads.

Reply:
@rob-solana since poh is a headless producer and there is no channel to the banking stage i've added an atomic to exit them both if either side exits